### ISSUE TITLE
`cargo clippy --all-targets` issues diagnostics for library crate twice

### ISSUE THREAD

**Issue by zackw** [rust-lang/rust-clippy#4092]
Given a crate containing this file as `src/lib.rs`

```rust
pub fn demo() -> &'static str {
    return "demo";
}
```

and at least one file in `tests` that does `use name_of_crate::demo`, `cargo clippy --all-targets` will diagnose the `needless_return` twice in a row.  If you use `cargo --verbose` it's clear that the clippy-driver program has been invoked twice (once with `--test` and once without) and each invocation is reporting the lint, but without that, it looks like clippy is repeating itself for no reason.

**Comment by GeeWee** [comment]
Can confirm that it also seems like `clippy fix` will attempt to diagnose and fix the problem multiple times within one invocation with `--all-targets`, (if you recreate the issue right after it was fixed)
